crates/bevy_math/src/curve/mod.rs

```diff
@@ -0,0 +1,610 @@
+    ///
+    /// Note that this is the opposite of what one might expect intuitively; for example, if this
+    /// curve has a parameter interval of `[0, 1]`, then linearly mapping the parameter domain to
+    /// `[0, 2]` would be performed as follows, dividing by what might be perceived as the scaling
+    /// factor rather than multiplying:
+    /// ```
+    /// # use bevy_math::curve::*;
+    /// let my_curve = constant_curve(interval(0.0, 1.0).unwrap(), 1.0);
+    /// let domain = my_curve.domain();
+    /// let scaled_curve = my_curve.reparametrize(interval(0.0, 2.0).unwrap(), |t| t / 2.0);
+    /// ```
+    /// This kind of linear remapping is provided by the convenience method
+    /// [`Curve::reparametrize_linear`], which requires only the desired domain for the new curve.
+    ///
+    /// # Examples
+    /// ```
+    /// // Reverse a curve:
+    /// # use bevy_math::curve::*;
+    /// # use bevy_math::vec2;
+    /// let my_curve = constant_curve(interval(0.0, 1.0).unwrap(), 1.0);
+    /// let domain = my_curve.domain();
+    /// let reversed_curve = my_curve.reparametrize(domain, |t| domain.end() - t);
+    ///
+    /// // Take a segment of a curve:
+    /// # let my_curve = constant_curve(interval(0.0, 1.0).unwrap(), 1.0);
+    /// let curve_segment = my_curve.reparametrize(interval(0.0, 0.5).unwrap(), |t| 0.5 + t);
+    ///
+    /// // Reparametrize by an easing curve:
+    /// # let my_curve = constant_curve(interval(0.0, 1.0).unwrap(), 1.0);
+    /// # let easing_curve = constant_curve(interval(0.0, 1.0).unwrap(), vec2(1.0, 1.0));
+    /// let domain = my_curve.domain();
+    /// let eased_curve = my_curve.reparametrize(domain, |t| easing_curve.sample(t).y);
+    /// ```
+    fn reparametrize<F>(self, domain: Interval, f: F) -> ReparamCurve<T, Self, F>
+    where
+        Self: Sized,
+        F: Fn(f32) -> f32,
+    {
+        ReparamCurve {
+            domain,
+            base: self,
+            f,
+            _phantom: PhantomData,
+        }
+    }
+
+    /// Linearly reparametrize this [`Curve`], producing a new curve whose domain is the given
+    /// `domain` instead of the current one. This operation is only valid for curves with finite
+    /// domains; if either this curve's domain or the given `domain` is infinite, an
+    /// [`InfiniteIntervalError`] is returned.
+    fn reparametrize_linear(
+        self,
+        domain: Interval,
+    ) -> Result<LinearReparamCurve<T, Self>, InfiniteIntervalError>
+    where
+        Self: Sized,
+    {
+        if !domain.is_finite() {
+            return Err(InfiniteIntervalError);
+        }
+
+        Ok(LinearReparamCurve {
+            base: self,
+            new_domain: domain,
+            _phantom: PhantomData,
+        })
+    }
+
+    /// Reparametrize this [`Curve`] by sampling from another curve.
+    ///
+    /// TODO: Figure out what the right signature for this is; currently, this is less flexible than
```

Comment:
Swap this to a normal comment: I'm fine to leave this in but it shouldn't show up in our docs.

Reply:
Good find 😅 I think that me-of-the-past was actually a bit misguided, and this should take ownership of `C`, since otherwise the resulting curve has a lifetime bounded by that of `C`. I think that's a bigger footgun than taking ownership of `C` when you didn't mean to, so I went ahead and made that change.